zqd/api/connection.go

```diff
@@ -0,0 +1,364 @@
+func (c *Connection) stream(ctx context.Context, method, path string, hdr http.Header, body interface{}) (*bufio.Scanner, error) {
+	resp, err := c._call(ctx, method, path, hdr, body)
+	if err != nil {
+		return nil, err
+	}
+	return NewJSONPipeScanner(resp.Body), nil
+}
+
+type ErrorResponse struct {
+	*http.Response
+	Err error
+}
+
+func responseError(resp *http.Response) error {
+	if resp.StatusCode < 400 {
+		return nil
+	}
+	resErr := &ErrorResponse{Response: resp}
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+	if strings.HasPrefix(resp.Header.Get("Content-Type"), "application/json") {
+		var apierr *Error
+		err := json.Unmarshal(body, apierr)
+		if err != nil {
+			resErr.Err = err
+		} else {
+			resErr.Err = apierr
+		}
+	} else {
+		resErr.Err = errors.New(string(body))
+	}
+	return resErr
+}
+
+func (e *ErrorResponse) Unwrap() error {
+	return e.Err
+}
+
+func (e *ErrorResponse) Error() string {
+	return fmt.Sprintf("status code %d: %v", e.StatusCode, e.Err)
+}
+
+// Ping checks to see if the server and measure the time it takes to
+// get back the response.
+func (c *Connection) Ping() (time.Duration, error) {
+	now := time.Now()
+	if _, err := c.call("GET", "/status", nil, nil, nil); err != nil {
+		return 0, err
+	}
+	return time.Since(now), nil
+}
+
+// SpaceInfo retrieves information about the specified space.
+func (c *Connection) SpaceInfo(spaceName string) (*SpaceInfo, error) {
+	path := path.Join("/space", url.PathEscape(spaceName))
+	var res SpaceInfo
+	if status, err := c.call("GET", path, nil, nil, &res); err != nil {
+		if status == http.StatusNotFound {
+			return nil, ErrSpaceNotFound
+		}
+		return nil, err
+	}
+	return &res, nil
+}
+
+func (c *Connection) SpacePost(req SpacePostRequest) (*SpacePostResponse, error) {
+	var res SpacePostResponse
+	if status, err := c.call("POST", "/space", nil, req, &res); err != nil {
+		if status == http.StatusConflict {
+			return nil, ErrSpaceExists
+		}
+		return nil, err
+	}
+	return &res, nil
+}
+
+func (c *Connection) SpaceList() ([]string, error) {
+	var res []string
+	_, err := c.call("GET", "/space", nil, nil, &res)
+	return res, err
+}
+
+func (c *Connection) SpaceDelete(spaceName string) (err error) {
+	path := path.Join("/space", url.PathEscape(spaceName))
+	_, err = c.call("DELETE", path, nil, nil, nil)
+	return err
+}
+
+// Search sends a search task to the server and returns a Search interface
+// that the caller uses to stream back results via the Pull method.
+func (c *Connection) Search(req SearchRequest, format string, params url.Values) (Search, error) {
+	return c.SearchWithContext(context.Background(), req, format, params)
+}
+
+// SearchWithContext is like PostSearch, except that it takes a
+// context that will be used in the underlying http request.
+func (c *Connection) SearchWithContext(ctx context.Context, req SearchRequest, format string, params url.Values) (Search, error) {
```

Comment:
I would just make this call `Search` and drop the other function that doesn't take a context.

Reply:
ok

Reply:
done